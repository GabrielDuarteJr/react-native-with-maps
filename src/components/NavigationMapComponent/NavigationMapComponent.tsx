import React, {useMemo} from 'react';
import {GOOGLE_MAPS_APIKEY} from 'react-native-dotenv';
import MapView, {PROVIDER_GOOGLE, Marker} from 'react-native-maps';
import MapViewDirections from 'react-native-maps-directions';
import PinIcon from '~/assets/icons/Pin.svg';
import CarIcon from '~/assets/icons/Car.svg';
import colors from '~/utils/colors';
import {initialRegion} from '~/utils/constants';
import {mapStyle} from './styles';

interface NavigationMapInterface {
  origin: Cordinates;
  destination: Cordinates;
  carOrientation: number;
}

const NavigationMapComponent: React.FunctionComponent<
  NavigationMapInterface
> = ({origin, destination, carOrientation}) => {
  const originLoc = useMemo<Cordinates>(
    () => ({
      latitude: origin.latitude,
      longitude: origin.longitude,
    }),
    [origin],
  );

  const destinationLoc = useMemo<Cordinates>(
    () => ({
      latitude: destination.latitude,
      longitude: destination.longitude,
    }),
    [destination],
  );

  return (
    <MapView
      style={mapStyle}
      provider={PROVIDER_GOOGLE}
      initialRegion={initialRegion}
      showsPointsOfInterest={false}
      showsIndoors={false}
      showsUserLocation>
      <MapViewDirections
        origin={originLoc}
        destination={destinationLoc}
        apikey={GOOGLE_MAPS_APIKEY}
        strokeWidth={6}
        strokeColor={colors.blue}
        mode="DRIVING"
        resetOnChange={false}
      />
      <Marker
        coordinate={origin}
        title={'Origin'}
        anchor={{x: 0.5, y: 0.5}}
        rotation={carOrientation}
        flat>
        <CarIcon />
      </Marker>
      <Marker
        coordinate={destination}
        title={'Destination'}
        anchor={{x: 0.5, y: 0.75}}
        flat>
        <PinIcon />
      </Marker>
    </MapView>
  );
};

export default NavigationMapComponent;
