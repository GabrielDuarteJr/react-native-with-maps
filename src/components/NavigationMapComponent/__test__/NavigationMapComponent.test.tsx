import 'react-native';
import React from 'react';
import {render, cleanup, screen} from '@testing-library/react-native';
import NavigationMapComponent from '../NavigationMapComponent';

jest.mock('react-native-maps', () => {
  const {View, Text} = require('react-native');
  const MockMapView = (props: any) => {
    return <View>{props.children}</View>;
  };
  const MockMarker = (props: any) => {
    return (
      <View>
        <Text>{props.coordinate.toString()}</Text>
        <Text>{props.rotation}</Text>
        {props.children}
      </View>
    );
  };
  return {
    __esModule: true,
    default: MockMapView,
    Marker: MockMarker,
  };
});

jest.mock('react-native-maps-directions', () => {
  const {View} = require('react-native');
  const MockMapDirectionsView = (props: any) => {
    return <View>{props.children}</View>;
  };
  return {
    __esModule: true,
    default: MockMapDirectionsView,
  };
});

afterEach(cleanup);

describe('NavigationMapComponent', () => {
  it('should render without crashing', () => {
    const cordinatesMock: Cordinates = {
      latitude: -18.914931,
      longitude: -48.244491,
    };

    const tree = render(
      <NavigationMapComponent
        origin={cordinatesMock}
        destination={cordinatesMock}
        carOrientation={20}
      />,
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should display the cordinates correctly', () => {
    const cordinatesMock: Cordinates = {
      latitude: -18.914931,
      longitude: -48.244491,
    };

    render(
      <NavigationMapComponent
        origin={cordinatesMock}
        destination={cordinatesMock}
        carOrientation={20}
      />,
    );

    const cordinates = screen.getAllByText(cordinatesMock.toString());

    expect(cordinates[0]).toBeDefined();
  });

  it('should display the rotation correctly', () => {
    const cordinatesMock: Cordinates = {
      latitude: -18.914931,
      longitude: -48.244491,
    };

    render(
      <NavigationMapComponent
        origin={cordinatesMock}
        destination={cordinatesMock}
        carOrientation={20}
      />,
    );

    const rotation = screen.getByText('20');

    expect(rotation).toBeDefined();
  });
});
