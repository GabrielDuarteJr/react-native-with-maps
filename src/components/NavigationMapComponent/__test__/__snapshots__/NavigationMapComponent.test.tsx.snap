// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationMapComponent should render without crashing 1`] = `
<View>
  <View />
  <View>
    <Text>
      [object Object]
    </Text>
    <Text>
      20
    </Text>
    <SvgMock />
  </View>
  <View>
    <Text>
      [object Object]
    </Text>
    <Text />
    <SvgMock />
  </View>
</View>
`;
